distributed/worker_memory.py

```diff
@@ -344,21 +345,40 @@ def memory_monitor(self, nanny: Nanny) -> None:
         except (ProcessLookupError, psutil.NoSuchProcess, psutil.AccessDenied):
             return  # pragma: nocover
 
-        if process.pid in (self._last_terminated_pid, None):
-            # We already sent SIGTERM to the worker, but its handler is still running
-            # since the previous iteration of the memory_monitor - for example, it
-            # may be taking a long time deleting all the spilled data from disk.
+        if memory / self.memory_limit <= self.memory_terminate_fraction:
             return
-        self._last_terminated_pid = -1
 
-        if memory / self.memory_limit > self.memory_terminate_fraction:
+        if self._last_terminated_pid != process.pid:
```

Comment:
The default interval is 100ms. Isn't this a bit short for us to escalate to kill?

Reply:
Not really - that's 100ms worth of extra leaking from any task that is currently running.

Note that, to the best of my understanding, no cleanup code whatsoever runs on SIGTERM by default. So it should be immediate, unless the user tampered with the signal handlers - which they could legitimately do for their own cleanup code. A realistic use case is a database library that installs a SIGTERM handler to cleanly shut down its sockets.

In this case, I argue that prompt termination for a process that is close to go beyond 100% is more important the clean teardown.

Reply:
We do install our own signal handlers and close workers/nannies, e.g.

https://github.com/dask/distributed/blob/7665eaacf0780594c33670e75b3517380102263f/distributed/_signals.py#L11

https://github.com/dask/distributed/blob/7665eaacf0780594c33670e75b3517380102263f/distributed/cli/dask_worker.py#L479-L487

Reply:
I know. This doesn't exclude third party handlers.
Typical pattern:
```python

def my_handler(signo, frame):
    # TODO do your thing
    prev(signo, frame)

prev = signal.signal(signal.SIGTERM, my_handler)
```